### ISSUE TITLE
Next word prediction - Presage

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#69]
After each letter when spelling a word - create word completions that are tolerant of misspellings AND whole word corrections (assuming word is complete but incorrectly spelled)

  Multi key capture

```
Netspell

Nhunspell
```

**Comment by JuliusSweetland** [comment]
Could be solved as part of the potential integration of Presage for next word prediction.

**Comment by JuliusSweetland** [comment]
From Arnaud (https://github.com/amartel/OptiKey)

Install and setup PRESAGE is very easy, especially for you as english is the default language database (English language is generated from single novel "The picture of Dorian Gray" by Oscar Wilde)
1) download the installer for Windows 32 bits (and not the 64bits version) at http://sourceforge.net/projects/presage/files/presage/0.9.1/win32/presage-0.9.1-32bit-setup.exe/download, run it and install all the options (wcf service is required and will start automatically at user login).
2) in Optikey project, add a reference to "C:\Program Files (x86)\presage\bin\presage.net.dll ». 
3) Sample code to integrate PRESAGE

using presage;

//context to be used to predict the next suggestion. For Optikey, this can be all the text entered in the scratchpad area, including in progress word (using private variable Text in KeyboardOutputService)
string Text;

//callback used to return the context for the predictions
private string callback_get_past_stream() {
	return Text;
}

//I don’t really understand the use of this callback function but, when looking in the presage samples, it always return an empty string ;-)
private string callback_get_future_stream() {
	return "";
}

Presage prsg = new Presage(this.callback_get_past_stream, this.callback_get_future_stream);
string[] suggestions = prsg.predict();

Note: I had crashes when using a single instance of Presage object. The workaround was to instantiate a new Presage object each time I wanted suggestions.

4) To build a new language model for presage:
a) export all the text you want to use in a text file with encoding ISO-8859-1 (for example, export.txt). Note: For my friend, Cédric, I exported all the mails in my OUTBOX folder
b) generate the model with the following commands:
 "C:\Program Files (x86)\presage\bin\text2ngram.exe"  -n 1 -f sqlite -o my_model.db export.txt
 "C:\Program Files (x86)\presage\bin\text2ngram.exe"  -n 2 -f sqlite -o my_model.db export.txt
 "C:\Program Files (x86)\presage\bin\text2ngram.exe"  -n 3 -f sqlite -o my_model.db export.txt
c) edit file "C:\Program Files (x86)\presage\etc\presage.xml" to change the database name
<DBFILENAME>[full path of the my_model.db file]</DBFILENAME>

You can also copy the original presage.xml file in C:\Users\[user]\.presage\presage.xml then modify it to reference your own language model (in this case, you don’t need to have an administrator access for the current user session)

d) That’s all!!! Presage will now use autolearning with all texts in the scratchpad and will build an additional database saved in the C:\Users\[user]\.presage directory

**Comment by JuliusSweetland** [comment]
A big concern is how do we bundle the install and startup/exit of Presage with OptiKey install, startup, and exit? I wouldn't want users to have to install and configure Presage to use it. Ideally OptiKey would include all language specific config which it would send to Presage on startup for the current language.

**Comment by JuliusSweetland** [comment]
Release v2.3.0 adds correction for spelling mistakes when generating autocomplete suggestions. This issue has been edited to next word prediction only.

**Comment by mostlyjason** [comment]
This would make text entry much faster! I looked at how other keyboard apps do it. It looks like acat has the user install presage https://01.org/sites/default/files/documentation/acat_user_guide_6.pdf. Another option could be using SwiftKeys prediction engine. It looks like they made it available free to Click2Speak. I could help ask them if you're interested. http://www.click2speak.net

**Comment by JuliusSweetland** [comment]
@mostlyjason Hi Jason. Yes ACAT uses Presage, which you have to train for each language. I also spoke to SwiftKey about using their API, but they had closed the API when I spoke to them so that was a bit of a dead end.

If you'd like to work on either solution I'd be happy to support.

**Comment by mostlyjason** [comment]
I might be able to help out I'm not sure how quick I'll be since I'm not familiar with the code for this program. Do you know where Arnaud added the Presage code, like what file or function? If not, do you have a suggestion or pointer?

**Comment by mostlyjason** [comment]
So I found the commit where he added presage https://github.com/amartel/OptiKey/commit/265a481e5ea586cf60c3bd2e35af20e4fbbd33ea. I'll try to run it to see if it works and possibly might be able to cherry pick some of the code into the latest version.

**Comment by JuliusSweetland** [comment]
Excellent. Looking forward to your next update!

**Comment by alexandre-mbm** [comment]
> Yes ACAT uses Presage, which you have to train for each language.

Could we get all from ACAT?

**Comment by mostlyjason** [comment]
@alexandre-mbm I think there is another issue that's been opened to discuss the integration of ACAT https://github.com/OptiKey/OptiKey/issues/163. I'm afraid a project like that goes beyond my development skills though. Do you have an idea how that would work?

**Comment by alexandre-mbm** [comment]
@mostlyjason, I suppose that you are speaking about the word prediction. It interest me. But I have no experience in fact. I would study. I'm lightly introduced into AI. But I think everything is already implemented. In case of the licenses permit us, maybe we just need get data (training results) from ACAT, not?

**Comment by mostlyjason** [comment]
Update: I got @amartel's code working with presage and it seems to be working well. @alexandre-mbm I think for this issue I just want to focus on the next word prediction which I think I can integrate fairly easy using Arnoud's example. Let's discuss integrating ACAT as a separate issue since I think it will be more complex.

As a start, I'd like to make a "beta" version with support for presage. This will include the default English dictionary that presage uses and will require the user to manually install presage in a separate step. Presage will be off by default but they can enable it by editing the settings file. Because it's off by default it can be merged into the master branch. This version could be used by more advanced English-language users. 

For the next version, we can enhance it with support for the Optikey dictionary, additional languages, an integrated installation, and turn it on by default for all users. What you think about these phases @JuliusSweetland?

In terms of implementation, I cannot directly merge Arnoud's code because it was written before the auto complete code was added. I like the design pattern for the auto complete code because it allows users to switch the algorithms used. I plan to create a new instance of IManageAutoComplete called PresageAutoComplete and put the new code there. @JuliusSweetland do you think this is the right place?

**Comment by JuliusSweetland** [comment]
@mostlyjason That all sounds good, with the caveat that the auto complete is designed to complete the current, in-progress word. If it were extended to do this as well as supply the next word would the current workflow still make sense? Alternatively we could have a new INextWord service that is called once each word is complete and populates suggestions seperately, so IManageAutoComplete is queried for in-progress typing, and INextWord (or whatever you think this interface should be called) for next word when we are happy that no word is in-progress. What do you think?

**Comment by mostlyjason** [comment]
@JuliusSweetland think that's a good point!  Actually presage can handle both completing the current word and suggesting the next word. The interface doesn't distinguish between them, you just give it the previous text and it will offer suggestions for what comes next. I could modify the auto complete interface to provide suggestions for both the current word and the next word. Your current NGramAutoComplete would just return null if has no next word suggestion. Alternatively, I could rename IAutoComplete to ISuggestion if we think that naming would be more clear. Do you have any opinion? If not, I can just take a crack at it and you can review when I'm done.

**Comment by JuliusSweetland** [comment]
I like the idea of ISuggestion and for that to handle both completing words
and next word.

Would you like a new branch forked from master to work against?

**Comment by mostlyjason** [comment]
I was planning to create my own fork and then submit a pull request to your master branch. Does that sound like a good workflow or do you prefer me to work off your fork?

**Comment by alexandre-mbm** [comment]
@mostlyjason, I was not thinking in ACAT integration. This that you thinks already. I want to try it with pt-br. I do not realize because you could not to work a branch fork from master.

**Comment by mostlyjason** [comment]
Just submitted PR please review!

**Comment by JuliusSweetland** [comment]
Released in v2.7.0
